webauthn-server-core/src/main/java/com/yubico/webauthn/FinishAssertionSteps.java

```diff
@@ -124,88 +125,68 @@ public void validate() {
   @Value
   class Step6 implements Step<Step7> {
 
-    private final Optional<ByteArray> userHandle =
+    private final Optional<ByteArray> requestedUserHandle =
         OptionalUtil.orElseOptional(
             request.getUserHandle(),
-            () ->
-                OptionalUtil.orElseOptional(
-                    response.getResponse().getUserHandle(),
-                    () ->
-                        request
-                            .getUsername()
-                            .flatMap(credentialRepository::getUserHandleForUsername)));
+            () -> request.getUsername().flatMap(credentialRepository::getUserHandleForUsername));
+
+    private final Optional<ByteArray> assertedUserHandle = response.getResponse().getUserHandle();
+
+    private final Optional<ByteArray> userHandle =
+        OptionalUtil.orElseOptional(assertedUserHandle, () -> requestedUserHandle);
 
     private final Optional<String> username =
         OptionalUtil.orElseOptional(
             request.getUsername(),
-            () -> userHandle.flatMap(credentialRepository::getUsernameForUserHandle));
+            () -> assertedUserHandle.flatMap(credentialRepository::getUsernameForUserHandle));
+
+    private final Optional<UserIdentity> userIdentity =
+        username.flatMap(
+            un ->
+                assertedUserHandle.map(
+                    uh -> UserIdentity.builder().name(un).displayName(un).id(uh).build()));
```

Comment:
```
.name(un).displayName(un)
```
This is incorrect, and cannot be resolved without adding additional methods to `CredentialRepository` just for the sake of this feature. For this reason I don't think the `UserIdentity` type should be involved at all in authentication ceremonies - the library has no reason to need it, so it should not be required by the supported interfaces.